cmd/skopeo/sync.go

```diff
@@ -0,0 +1,594 @@
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"github.com/urfave/cli"
+)
+
+type syncOptions struct {
+	global            *globalOptions
+	srcImage          *imageOptions
+	destImage         *imageDestOptions
+	removeSignatures  bool   // Do not copy signatures from the source image
+	signByFingerprint string // Sign the image using a GPG key with the specified fingerprint
+	sourceYaml        bool
+}
+
+type repoDescriptor struct {
+	DirBasePath  string // base path when source is 'dir'
+	TaggedImages []types.ImageReference
+	Context      *types.SystemContext
+}
+
+type tlsVerifyConfig struct {
+	skip bool
+}
+
+type registrySyncCfg struct {
+	Images      map[string][]string
+	Credentials types.DockerAuthConfig
+	TLSVerify   tlsVerifyConfig `yaml:"tls-verify"`
+	CertDir     string          `yaml:"cert-dir"`
+}
+
+type sourceCfg map[string]registrySyncCfg
+
+// Custom unmarshaler for tls-verify, so when user does not specify the TLS
+// verify option, then it is ON by default.
+func (tls *tlsVerifyConfig) UnmarshalYAML(unmarshal func(interface{}) error) error {
+	var verify bool
+	if err := unmarshal(&verify); err != nil {
+		return err
+	}
+
+	tls.skip = !verify
+
+	return nil
+}
+
+// Generates a config structure from a YAML file.
+func newSourceConfig(yamlFile string) (cfg sourceCfg, err error) {
+	err = yamlUnmarshal(yamlFile, &cfg)
+	return
+}
+
+// Checks if a given transport is supported by the sync operation.
+func validSyncTransport(transport types.ImageTransport) bool {
+	switch transport {
+	case docker.Transport:
+		return true
+	case directory.Transport:
+		return true
+	}
+
+	return false
+}
+
+// Return a URL object from an input string
+func parseURL(urlString string) (*url.URL, error) {
+	var parsedURL *url.URL
+
+	parsedURL, err := url.Parse(urlString)
+	if err != nil {
+		return parsedURL, errors.WithMessage(err, "Error while parsing source")
+	}
+
+	valid := validSyncTransport(transports.Get(parsedURL.Scheme))
+	if !valid {
+		return parsedURL, errors.New("Invalid transport")
+	}
+
+	return parsedURL, nil
+}
+
+// Given a tranport and an image name (without the transport), returns an
+// ImageReference.
+func getImageReference(transport types.ImageTransport, imgName string) (types.ImageReference, error) {
+	ref, err := transport.ParseReference(imgName)
+	if err != nil {
+		return nil, errors.WithMessage(err, fmt.Sprintf("Cannot obtain a valid image reference from '%s'", imgName))
+	}
+
+	return ref, nil
+}
+
+// Given a directory as a URL object, returns its string representation suitable
+// to be used as a filesystem path
+func dirPathFromURL(dirURL *url.URL) (string, error) {
+	var dirPath string
+	if dirURL.Scheme != directory.Transport.Name() {
+		return "", fmt.Errorf("Not an dir URL: %v", dirURL)
+	}
+
+	if dirURL.Opaque != "" {
+		// relative dir path, e.g. dir:localdir (without "/" or "//")
+		dirPath = dirURL.Opaque
+	} else {
+		dirPath = path.Join("/", dirURL.Host, dirURL.Path)
+	}
```

Comment:
The _transport_`:`_image_ syntax really is not URLs (e.g. no percent-escaping) ; overall I don’t think it’s reasonable to mix the two.

I realize the code needs to detect whether a `docker://` reference was specified with a tag, and the current reference API makes that impossible, so _something_ needs to change (should the sources/destinations even look similar when the `:latest` tag defaulting is not there?); still, using the `URL` parser is not making that problem any easier AFAICT.

Reply:
I opted for `url.URL` because in multiple occasions I needed to extract {transport, registry hostname, image path} and generate new URLs with them.
Doing it using low level `string` required more code and replicating in part what `net/url` is already doing.

I did not thought about handling escaping correctly, thanks for raising this. Looks like you can deal with by using the value of [`URL.RawPath`](https://golang.org/pkg/net/url/#URL).

Overall I feel using URL is adequate to parsing _transport_`:`_image_, if you take into account things like escaping and `dir:` non-absolute paths.

Reply:
I share @mtrmac thoughts. `net/url.Parse(...)` is full of very odd behaviors. We tried fighting it multiple times and ended up avoiding it entirely.

Can't we use "github.com/containers/image/transports/alltransports" instead of using "net/url.URL"?

Reply:
> I opted for `url.URL` because in multiple occasions I needed to extract {transport, registry hostname, image path} and generate new URLs with them.

None of those values are URLs, using the proper definition.

> Doing it using low level `string` required more code and replicating in part what `net/url` is already doing.

Almost nothing `net/url` is doing is applicable or correct for those values. At best `net/url.getScheme` is a very complex way to do `strings.SplitN(… ":", 2)` — with extra logic that either does not matter or is incompatible with the `alltransports.ParseImageName` parser.

And this code ends up having to correct for the mismatches, like the `if` quoted above, or `fmt.Sprintf("//%s%s", sourceURL.Host, sourceURL.Path)` elsewhere.

The use of `url.URL` is _already_ adding more code than it saves, and that’s even before getting the corner cases correct.

(I do realize I was speaking against hand-rolled parsers in 
https://github.com/containers/skopeo/pull/524#discussion_r210649873 , but using a mismatching parser is not better.)

> Overall I feel using URL is adequate to parsing _transport_`:`_image_, if you take into account things like escaping and `dir:` non-absolute paths.

Well, but `dir:` _does not do escaping_ of any kind, so using a parser that does would interpret the input differently.

(Arguably `oci:` and `oci-archive:` _should_ do some escaping because the `oci:`_path_`:`_imageName` format can’t represent _path_ values that contain a `:`, but it’s quite unclear how to map that to an URL in a way that makes sense while still using the URL reserved characters vs. encoded characters to represent this.)